File: src/main/java/io/github/intellij/dlanguage/stubs/types/DeclaratorIdentifierStubElementType.java
package io.github.intellij.dlanguage.stubs.types;

import com.intellij.psi.PsiElement;
import com.intellij.psi.stubs.StubElement;
import com.intellij.psi.stubs.StubInputStream;
import com.intellij.psi.stubs.StubOutputStream;
import io.github.intellij.dlanguage.psi.named.DlangDeclaratorIdentifier;
import io.github.intellij.dlanguage.psi.impl.named.DlangDeclaratorIdentifierImpl;
import io.github.intellij.dlanguage.stubs.DlangDeclaratorIdentifierStub;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;

public class DeclaratorIdentifierStubElementType extends
    DNamedStubElementType<DlangDeclaratorIdentifierStub, DlangDeclaratorIdentifier> {
    public DeclaratorIdentifierStubElementType(String debugName) {
        super(debugName);
    }

    @Override
    public DlangDeclaratorIdentifier createPsi(@NotNull final DlangDeclaratorIdentifierStub stub) {
        return new DlangDeclaratorIdentifierImpl(stub, this);
    }

    @Override
    public @NotNull DlangDeclaratorIdentifierStub createStub(@NotNull DlangDeclaratorIdentifier psi, StubElement<? extends PsiElement> parentStub) {
        return new DlangDeclaratorIdentifierStub(parentStub, this, psi.getName(), null);
    }

    @Override
    public void serialize(@NotNull final DlangDeclaratorIdentifierStub stub, @NotNull final StubOutputStream dataStream) throws IOException {
        dataStream.writeName(stub.getName());
    }
    @Override
    public @NotNull DlangDeclaratorIdentifierStub deserialize(@NotNull StubInputStream dataStream, StubElement parentStub) throws IOException {
        return new DlangDeclaratorIdentifierStub(parentStub, this, dataStream.readName(), null);
    }
}

File: dlang/psi-impl/src/main/kotlin/io/github/intellij/dlanguage/stubs/types/DeclaratorIdentifierStubElementType.kt
package io.github.intellij.dlanguage.stubs.types

import com.intellij.psi.PsiElement
import com.intellij.psi.stubs.StubElement
import com.intellij.psi.stubs.StubInputStream
import com.intellij.psi.stubs.StubOutputStream
import io.github.intellij.dlanguage.psi.impl.named.DlangDeclaratorIdentifierImpl
import io.github.intellij.dlanguage.psi.named.DLanguageDeclaratorIdentifier
import io.github.intellij.dlanguage.stubs.DLanguageDeclaratorIdentifierStub
import java.io.IOException

class DeclaratorIdentifierStubElementType(debugName: String) :
    DNamedStubElementType<DLanguageDeclaratorIdentifierStub, DLanguageDeclaratorIdentifier>(debugName) {
    override fun createPsi(stub: DLanguageDeclaratorIdentifierStub): DLanguageDeclaratorIdentifier {
        return DlangDeclaratorIdentifierImpl(stub, this)
    }

    override fun createStub(
        psi: DLanguageDeclaratorIdentifier,
        parentStub: StubElement<out PsiElement?>?
    ): DLanguageDeclaratorIdentifierStub {
        return DLanguageDeclaratorIdentifierStub(parentStub, this, psi.name, null)
    }

    @Throws(IOException::class)
    override fun serialize(stub: DLanguageDeclaratorIdentifierStub, dataStream: StubOutputStream) {
        dataStream.writeName(stub.name)
    }

    @Throws(IOException::class)
    override fun deserialize(dataStream: StubInputStream, parentStub: StubElement<*>?): DLanguageDeclaratorIdentifierStub {
        return DLanguageDeclaratorIdentifierStub(parentStub, this, dataStream.readName(), null)
    }
}
